src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -0,0 +1,350 @@
+      // Make a list of counts of all UMIs in order from most to least abundant
+      val nodes = SimpleCounter(rawUmis).map(x => new Node(x._1, x._2)).toBuffer.sortBy((n:Node) => -n.count)
+      val roots = mutable.Buffer[Node]()
+
+      // Now build one or more graphs starting with the most abundant remaining umi
+      def addChildren(root: Node, others: mutable.Buffer[Node]) : Unit = {
+        val xs = others.filter(other => countMismatches(root.umi, other.umi) <= maxMismatches && root.count >= 2 * other.count - 1)
+        root.children ++= xs
+        others --= xs
+        root.children.foreach(r => addChildren(r, others))
+      }
+
+      while (nodes.nonEmpty) {
+        val root = nodes.remove(0)
+        addChildren(root, nodes)
+        roots += root
+      }
+
+      // And finally make the output mapping
+      val mappings = mutable.Buffer[(String,String)]()
+      roots.foreach(root => {
+        mappings.append((root.umi, root.umi))
+        root.children.foreach(child => mappings.append((child.umi, root.umi)))
+      })
+
+      mappings.toMap
+    }
+  }
+
+  /** Looks in all the places the library name can be hiding. Returns the library name
+    * if one is found, otherwise returns "unknown".
+    */
+  final def library(rec: SAMRecord): String = {
+    var lib: String = rec.getTransientAttribute(LibraryAttribute).asInstanceOf[String]
+    if (lib == null) {
+      val rg = rec.getReadGroup
+      if (rg != null && rg.getLibrary != null) lib = rg.getLibrary
+      else lib = rec.getStringAttribute("LB")
+
+      if (lib == null) lib = "unknown"
+      rec.setTransientAttribute(LibraryAttribute, lib)
+    }
+
+    lib
+  }
+
+  /** Comparator to sort the reads by the earlier of the unclipped 5' position of the
+    * first or second read.  Groups reads in a convenient way for duplicate-marking and
+    * UMI assignment, and ensures that both ends of a read come together.
+    *
+    * Private because it has some serious restrictions!  Only allows:
+    *   - Paired reads only
+    *   - Mapped reads with mapped mates only
+    *   - No secondary or supplementary reads
+    *   - Read and mate _must_ be mapped to the same chromosome
+    * If any of these conditions are violated it will go badly!
+    */
+  private[umi] class EarlierReadComparator extends SAMRecordComparator {
+    override def fileOrderCompare(lhs: SAMRecord, rhs: SAMRecord): Int = compare(lhs, rhs)
+
+    /** A set of functions that extract a value from a read, used in order to compare reads. */
+    private val comparisons: Seq[(SAMRecord,SAMRecord) => Int] = Seq(
+      (lhs, rhs) => lhs.getReferenceIndex - rhs.getReferenceIndex,
+      (lhs, rhs) => sortPosition(lhs) - sortPosition(rhs),
+      (lhs, rhs) => library(lhs).compareTo(library(rhs)),
+      (lhs, rhs) => lhs.getReadName.compareTo(rhs.getReadName),
+      (lhs, rhs) => if (lhs.getFirstOfPairFlag) -1 else 1,
+      (lhs, rhs) => lhs.getFlags - rhs.getFlags
```

Comment:
Admittedly it should never be reached, but for some reason I thought there should be a tie breaker in case duplicate reads got fed through.  I'll kill it until proved necessary.